TEST: IMPLEMENT MATERIAL TESTS

TomasCB24 opened the pull request:
This pull request focuses on the implementation of tests for the material-related endpoints in our application. The main objective is to guarantee the robustness and reliability of these functions, ensuring that they meet the functional requirements and do not present unexpected problems.

**For the reviewer**

Check if it complies with the established:

- [x] Tests have been added for all material-related endpoints.
- [x] Consistent mocks or test data have been used to simulate interactions with the database or other external services.
- [x] Tests are organised in a logical and coherent way.
- [ ] Structures such as test suites have been used to group related cases.
- [ ] Each test is independent and does not depend on the results of other tests.
- [ ] Tests are efficient and executed in a reasonable time.
- [ ] Tests run automatically as part of continuous integration (CI).
- [ ]  Tests follow project coding conventions and standards.

TomasCB24 committed:
TEST: IMPLEMENT MATERIAL TESTS

Co-authored-by: Miguel Romero Arjona <miguelromeroarjona@gmail.com>

TomasCB24 committed:
FIX: CHANGE TESTS.YML FILE

Co-authored-by: Miguel Romero Arjona <miguelromeroarjona@gmail.com>

TomasCB24 committed:
FIX: SET WORKING DIRECTORY

Co-authored-by: Miguel Romero Arjona <miguelromeroarjona@gmail.com>

TomasCB24 committed:
FIX: DELETE 14.X NODE VERSION

Co-authored-by: Miguel Romero Arjona <miguelromeroarjona@gmail.com>

TomasCB24 committed:
FIX: ADD JWT AS ENV VARIABLE

Co-authored-by: Miguel Romero Arjona <miguelromeroarjona@gmail.com>

TomasCB24 committed:
FIX: ADD JWT_EXPIRATION_TIME ENV VARIABLE

Co-authored-by: Miguel Romero Arjona <miguelromeroarjona@gmail.com>

ivamorgra commented on server/tests/materials.test.ts:
```diff
@@ -0,0 +1,640 @@
+    const createWriteStreamMock = jest.fn()
+    createWriteStreamMock.mockReturnValue({
+        on: jest.fn((event: string, callback: Function) => {
+            if (event === 'error') {
```
I would storage string variables in constants to avoid reapeat the same in multiple places

mrarjona replied:
> I would storage string variables in constants to avoid reapeat the same in multiple places

Ok, I will define constants to avoid duplicity of strings.

ivamorgra commented on server/tests/materials.test.ts:
```diff
@@ -0,0 +1,640 @@
+describe('Materials API', () => {
+    describe('GET /materials/:me', () => {
+        let findMaterialsByUsernameMock: jest.SpyInstance
+        let payload: IPayload = jwt.verify(JSON_WEB_TOKEN, JWT_SECRET, {
```
I would separate the verification of token from the unit test, althought this only exists here. This separation lets accomplish the principle of single responsability

mrarjona replied:
> I would separate the verification of token from the unit test, althought this only exists here. This separation lets accomplish the principle of single responsability

You are absolutely right. I will separate the verification logic from the test token.

ivamorgra commented on server/tests/materials.test.ts:
```diff
@@ -0,0 +1,640 @@
+                .set('Authorization', `Bearer ${JSON_WEB_TOKEN}`)
+
+            expect(response.status).toBe(200)
+            expect(response.body.length).toBe(1)
```
I think you could separate in diferent it to avoid the same test checks multiple responsabilities

mrarjona replied:
> I think you could separate in diferent it to avoid the same test checks multiple responsabilities

I think you are right about separating responsibilities, but I think that getting a correct answer also implies, in this case, that the body length is one. This avoids running two tests that make the same request and get the same result. Anyway we can discuss this at the next meeting.

ivamorgra reviewed (approved):
Great job overall! 🚀

Just a couple of suggestions to further enhance:

Avoid using 'any' and specify types:
Wherever possible, it would be great to replace the use of 'any' with specific types. This not only improves code safety but also makes it easier to understand.

Decouple configurations:
Consider decoupling configurations, such as URLs, tokens, and secrets, either through an external configuration file or a centralized object. This makes updates and maintenance in the future much more straightforward.

Thanks for your hard work! 👏

mrarjona commented:
> Great job overall! 🚀
> 
> Just a couple of suggestions to further enhance:
> 
> Avoid using 'any' and specify types: Wherever possible, it would be great to replace the use of 'any' with specific types. This not only improves code safety but also makes it easier to understand.
> 
> Decouple configurations: Consider decoupling configurations, such as URLs, tokens, and secrets, either through an external configuration file or a centralized object. This makes updates and maintenance in the future much more straightforward.
> 
> Thanks for your hard work! 👏

Thank you very much for the very complete review @ivamorgra! 💯  We take into account all your comments for the improvement of the developed code. 💪🏻

mrarjona committed:
FIX: IMPLEMENT SUGGESTIONS AFTER REVIEW OF THE PR

Constants have been added to eliminate duplication, any types have been eliminated and the principle of responsibility has been
taken into account, among other actions.

mrarjona committed:
Merge branch 'develop' into task/022

mrarjona committed:
FIX: RESOLVE BRANCH CONFLICTS
